src/main/java/com/shapesecurity/shift/es2016/codegen/TokenStream.java

```diff
@@ -67,17 +67,22 @@ public void put(@Nonnull String tokenStr) {
             if (String.valueOf(tokenStr).equals(".")) {
                 this.writer.append(numberNeedsDoubleDot(this.lastNumber) ? ".." : ".");
                 this.lastNumber = null;
-                this.lastChar = '.';
+                this.lastCodePoint = '.';
                 return;
             }
         }
         this.lastNumber = null;
-        char rightChar = tokenStr.charAt(0);
-        char lastChar = this.lastChar;
-        this.lastChar = tokenStr.charAt(tokenStr.length() - 1);
-        if ((lastChar == '+' || lastChar == '-') && lastChar == rightChar ||
-                Utils.isIdentifierPart(lastChar) && Utils.isIdentifierPart(rightChar) ||
-                lastChar == '/' && (rightChar == 'i' || rightChar == '/')) {
+        int rightCodePoint = tokenStr.codePointAt(0);
+        int lastCodePoint = this.lastCodePoint;
+        char lastChar = tokenStr.charAt(tokenStr.length() - 1);
+        if (lastChar >= 0xDC00 && lastChar <= 0xDFFF) {
+            this.lastCodePoint = tokenStr.codePointAt(tokenStr.length() - 2);
+        } else {
+            this.lastCodePoint = lastChar;
+        }
+        if ((lastCodePoint == '+' || lastCodePoint == '-') && lastCodePoint == rightCodePoint ||
```

Comment:
`lastCodePoint` is an integer, let's do integer comparisons. You can list the code point info in a comment above.

```java
// U+002B PLUS SIGN
// U+002D HYPHEN-MINUS
if ((lastCodePoint == 0x2B || lastCodePoint == 0x2D) && lastCodePoint == rightCodePoint ||
```

alternatively, we could use constants

```java
static final PLUS_SIGN = 0x2B; // U+002B PLUS SIGN
static final HYPHEN_MINUS = 0x2D; // U+002D HYPHEN-MINUS
// ...
if ((lastCodePoint == PLUS_SIGN || lastCodePoint == HYPHEN_MINUS) && lastCodePoint == rightCodePoint ||
```